R/PipeOpTorchFTTransformerBlock.R

```diff
@@ -0,0 +1,335 @@
+#' @title Single Transformer Block for FT-Transformer
+#' @description
+#' A transformer block, consisting of a multi-head self-attention mechanism followed by a feed-forward
+#' network
+#'
+#' This is used in the FT-Transformer.
+#' 
+#' TODO: verify that all of the assertions are present
+#' (although they will be factored out once we implement the learner)
+#'
+#' TODO: verify all documentation (LLM-generated)
+#' @param d_token (`integer(1)`)\cr
+#'   The dimension of the embedding.
+#' @param attention_n_heads (`integer(1)`)\cr
+#'   Number of attention heads.
+#' @param attention_dropout (`numeric(1)`)\cr
+#'   Dropout probability in the attention mechanism.
+#' @param attention_initialization (`character(1)`)\cr
+#'   Initialization method for attention weights. Either "kaiming" or "xavier".
+#' @param ffn_d_hidden (`integer(1)`)\cr
+#'   Hidden dimension of the feed-forward network.
+#' @param ffn_dropout (`numeric(1)`)\cr
+#'   Dropout probability in the feed-forward network.
+#' @param ffn_activation (`nn_module`)\cr
+#'   Activation function for the feed-forward network. Default value is `nn_reglu`.
+#' @param residual_dropout (`numeric(1)`)\cr
+#'   Dropout probability for residual connections.
+#' @param prenormalization (`logical(1)`)\cr
+#'   Whether to apply normalization before attention and FFN (TRUE) or after (FALSE). When this is FALSE, `first_prenormalization` must also be FALSE.
+#' @param is_first_layer (`logical(1)`)\cr
+#'   Whether this is the first layer in the transformer stack. Default value is FALSE.
+#' @param first_prenormalization (`logical(1)`)\cr
+#'   Whether to apply prenormalization in the first layer. It is recommended to set this to FALSE.
+#' @param attention_normalization (`function`)\cr
+#'   Normalization function to use for attention. Default value is `nn_layer_norm`.
+#' @param ffn_normalization (`function`)\cr
+#'   Normalization function to use for the feed-forward network. Default value is `nn_layer_norm`.
+#' @param kv_compression_ratio (`numeric(1)` or `NULL`)\cr
+#'   Ratio for key-value compression. If NULL, no compression is applied.
+#' @param kv_compression_sharing (`character(1)` or `NULL`)\cr
+#'   How to share compression weights. Options: "headwise", "key_value", or "layerwise".
+#' @param n_tokens (`integer(1)` or `NULL`)\cr
+#'   Number of tokens in the input sequence.
+#' @param query_idx (`integer()` or `NULL`)\cr
+#'   Indices to select for the query.
+#' @param attention_bias (`logical(1)`)\cr
+#'   Whether attention is biased. Default is TRUE.
+#' @param ffn_bias_first (`logical(1)`)\cr
+#'   Whether the first layer in the FFN has a bias. Default is TRUE.
+#' @param ffn_bias_second (`logical(1)`)\cr
+#'   Whether the second layer in the FFN has a bias. Default is TRUE.
+#'
+#' @references
+#' `r format_bib("devlin2018bert")`
+#'
+#' @export
+nn_ft_transformer_block = nn_module(
+  "nn_ft_transformer_block",
```

Comment:
we still need to specify the default values for the transformer layer here (for those that have one)